# Course: Application of AI, Data Science and Machine Learning
# Lab 4: Recommender System

## In this lab, we will learn to build recommeder system.

### Q1. Compute cosine similiarity between these 3 documents? Use term frequency to represent the documents and remove stop words before computing the similarity. Write your interpretation about which two documents are more similar? 
corpus=[ 'Julie loves me more than Linda loves me',
         'Jane likes me more than Julie loves me', 
         'harry likes kiwi fruit']

#### You can see that document 1 and 2 are more similar. You will also notice that their is differnce between the similarity score given by the cosine_similarity() and linear_kernel(). This happens becuase byy default  'CountVectorizer' fit.transform function does not length normalizes the data.

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
corpus=[ 'Julie loves me more than Linda loves me',
'Jane likes me more than Julie loves me',
'harry likes kiwi fruit'] 
# Creating coun vectorizer object. Note: we have removed the stop words
vectorizer = CountVectorizer(stop_words='english')
vectors = vectorizer.fit_transform(corpus)
print('\n ---Corpus converted to term-frequency vector:--\n',
vectors.toarray())
print('\n ---Unique terms in the corpus:--- \n',
vectorizer.get_feature_names())

# Import linear_kernel
from sklearn.metrics.pairwise import linear_kernel
from sklearn.metrics.pairwise import cosine_similarity
# Compute the cosine similarity matrix
cosine_sim = linear_kernel(vectors, vectors)
cosine_sim2 = cosine_similarity(vectors, vectors)

print(cosine_sim)
print(cosine_sim2)


 ---Corpus converted to term-frequency vector:--
 [[0 0 0 1 0 0 1 2]
 [0 0 1 1 0 1 0 1]
 [1 1 0 0 1 1 0 0]]

 ---Unique terms in the corpus:--- 
 ['fruit', 'harry', 'jane', 'julie', 'kiwi', 'likes', 'linda', 'loves']
[[6. 3. 0.]
 [3. 4. 1.]
 [0. 1. 4.]]
[[1.         0.61237244 0.        ]
 [0.61237244 1.         0.25      ]
 [0.         0.25       1.        ]]


/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


## Q2. Repeat the above excercise but this time use tf-idf weights to convert the corpus? 

#### You can see that document 1 and 2 are more similar. You will also notice that their is differnce between the similarity score given by the cosine_similarity() and linear_kernel(). This happens becuase byy default  'TfidfVectorizer()' fit.transform function length normalizes the data.

In [ ]:
#Import TfIdfVectorizer from scikit-learn
from sklearn.feature_extraction.text import TfidfVectorizer

corpus=[ 'Julie loves me more than Linda loves me',
'Jane likes me more than Julie loves me',
'harry likes kiwi fruit']
#Define a TF-IDF Vectorizer Object.
tfidf = TfidfVectorizer(stop_words='english')
#Construct the required TF-IDF matrix by fitting and transforming the data
tfidf_matrix = tfidf.fit_transform(corpus)

# Import linear_kernel
from sklearn.metrics.pairwise import linear_kernel
from sklearn.metrics.pairwise import cosine_similarity
# Compute the cosine similarity matrix
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)
cosine_sim2 = cosine_similarity(tfidf_matrix, tfidf_matrix)

print(cosine_sim)
print(cosine_sim2)

[[1.         0.53182464 0.        ]
 [0.53182464 1.         0.18487962]
 [0.         0.18487962 1.        ]]
[[1.         0.53182464 0.        ]
 [0.53182464 1.         0.18487962]
 [0.         0.18487962 1.        ]]


## Q3. Now build a complete recommeder system for a real world datatset. Read the data it the file 'movies_metadata.csv'. Use 'overview' attribute to compute similarity between the movies. Finally list top 10 similar movies to 'Father of the Bride Part II'. Note: You can use some basic pre-processing techniques for example, removing rows with missing values etc.

## Code credit
 #https://www.datacamp.com/community/tutorials/recommender-systems-python
## More on Text similiarity
https://medium.com/@adriensieg/text-similarities-da019229c894
## More on recommender system
https://goodboychan.github.io/chans_jupyter/python/datacamp/natural_language_processing/2020/07/17/04-TF-IDF-and-similarity-scores.html#Cosine-similarity

In [ ]:
# Import Pandas
import pandas as pd

#Import TfIdfVectorizer from scikit-learn
from sklearn.feature_extraction.text import TfidfVectorizer

# Import linear_kernel
from sklearn.metrics.pairwise import linear_kernel

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import sys
sys.path.insert(0,'/content/drive/My Drive/Colab Notebooks')

In [ ]:
# Load Movies Metadata
metadata = pd.read_csv('/content/drive/My Drive/Colab Notebooks/movies_metadata.csv', low_memory=False)

# Print the first three rows
print(metadata.head(3))


#Print plot overviews of the first 5 movies.
#print(metadata['overview'].head())
metadata['overview']



   adult  ... vote_count
0  False  ...     5415.0
1  False  ...     2413.0
2  False  ...       92.0

[3 rows x 24 columns]


0        Led by Woody, Andy's toys live happily in his ...
1        When siblings Judy and Peter discover an encha...
2        A family wedding reignites the ancient feud be...
3        Cheated on, mistreated and stepped on, the wom...
4        Just when George Banks has recovered from his ...
                               ...                        
45461          Rising and falling between a man and woman.
45462    An artist struggles to finish his work while a...
45463    When one of her hits goes wrong, a professiona...
45464    In a small town live two brothers, one a minis...
45465    50 years after decriminalisation of homosexual...
Name: overview, Length: 45466, dtype: object

In [ ]:
#Replace NaN with an empty string
metadata['overview'] = metadata['overview'].fillna('')

In [ ]:
#Define a TF-IDF Vectorizer Object. Remove all english stop words such as 'the', 'a'
tfidf = TfidfVectorizer(stop_words='english')



#Construct the required TF-IDF matrix by fitting and transforming the data
tfidf_matrix = tfidf.fit_transform(metadata['overview'])

#Output the shape of tfidf_matrix
print(tfidf_matrix.shape)

#Array mapping from feature integer indices to feature name.
print(tfidf.get_feature_names()[5000:5010])

(45466, 75827)
['avails', 'avaks', 'avalanche', 'avalanches', 'avallone', 'avalon', 'avant', 'avanthika', 'avanti', 'avaracious']


/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


In [ ]:
# Compute the cosine similarity matrix
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)

print('cosine similiarity matrix shape:', cosine_sim.shape)

#Construct a reverse map of indices and movie titles
indices = pd.Series(metadata.index, index=metadata['title']).drop_duplicates()

print(indices[:10])

cosine similiarity matrix shape: (45466, 45466)

title

Toy Story                      0

Jumanji                        1

Grumpier Old Men               2

Waiting to Exhale              3

Father of the Bride Part II    4

Heat                           5

Sabrina                        6

Tom and Huck                   7

Sudden Death                   8

GoldenEye                      9

dtype: int64

---



In [ ]:
# Function that takes in movie title as input and outputs most similar movies
def get_recommendations(title, cosine_sim=cosine_sim):
    # Get the index of the movie that matches the title
    idx = indices[title]

    # Get the pairwsie similarity scores of all movies with that movie
    sim_scores = list(enumerate(cosine_sim[idx]))

    # Sort the movies based on the similarity scores
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the scores of the 10 most similar movies
    sim_scores = sim_scores[1:11]

    # Get the movie indices
    movie_indices = [i[0] for i in sim_scores]

    # Return the top 10 most similar movies
    return metadata['title'].iloc[movie_indices]
get_recommendations('Father of the Bride Part II')

6793        Father of the Bride

6571                      Kuffs

6306            North to Alaska

19801                   Babbitt

34466         You're Killing Me

13611       The Magic of Méliès

5005                    Wendigo

27974          I Start Counting

43887    George of the Jungle 2

7097         The Out of Towners

Name: title, dtype: object

Code credit
https://www.datacamp.com/community/tutorials/recommender-systems-python

More on Text similiarity
https://medium.com/@adriensieg/text-similarities-da019229c894

More on recommender system
https://goodboychan.github.io/chans_jupyter/python/datacamp/natural_language_processing/2020/07/17/04-TF-IDF-and-similarity-scores.html#Cosine-similarity